### ISSUE TITLE
[FSDP] Support param step with fp32

### ISSUE THREAD

**Issue by yikaizhu-baseten** [pytorch/pytorch#167291]
### 🚀 The feature, motivation and pitch

In Megatron, we can keep a fp32 version of params. while doing optimizer.step, the gradient is used to update the fp32 version of params, and the cast the fp32 param to fp16 version. Can we do this in FSDP?

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

**Comment by weifengpy** [comment]
I believe you are talking about this init flow: single-device fp32 model -> fully_shard(mp_policy=(param_dtype=bf16, reduce_dtype=fp32)). That gives us opt.step on fp32 params + fp32 grads. example

https://github.com/pytorch/pytorch/blob/8f5c518dd325a7779d194c8961f670befb28f5d9/test/distributed/_composable/fsdp/test_fully_shard_mixed_precision.py#L195-L204
